### PR TITLE
Untyped error checks in catch statement

### PR DESCRIPTION
**Pull request by dirtydanee**
Implements https://github.com/realm/SwiftLint/issues/2045

### CODE REVIEW

**Comment by ornithocoder** [comment]
Don't forget to add your entry to the change log  :-)

File: Source/SwiftLintFramework/Rules/UntypedErrorRule.swift [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+//  SwiftLint
+//
+//  Created by Daniel.Metzing on 17/02/18.
+//  Copyright © 2018 Realm. All rights reserved.
+//
+
+import Foundation
+import SourceKittenFramework
+
+public struct UntypedErrorRule: ASTRule, OptInRule, ConfigurationProviderRule {
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/UntypedErrorRule.swift]: What do you think about changing it to `UntypedErrorInCatchRule`

Comment (marcelofabri) [reply Source/SwiftLintFramework/Rules/UntypedErrorRule.swift]: And then chaning the rule identifier and name as well

File: Source/SwiftLintFramework/Rules/UntypedErrorRule.swift [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        "\\{"     // Start scope character
+    )
+
+    public static let description = RuleDescription(
+        identifier: "untyped_error",
+        name: "Untyped error rule",
+        description: "Catch statements should not declare error variables without type casting",
+        kind: .lint,
+        nonTriggeringExamples: [
+            "do {\n try foo() \n} catch {\n}",
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/UntypedErrorRule.swift]: can you format he entries better? Check how they're rendered in Rules.md. I think just improving the indentation of `try foo()` would already help.

File: Source/SwiftLintFramework/Rules/UntypedErrorRule.swift [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+public struct UntypedErrorRule: ASTRule, OptInRule, ConfigurationProviderRule {
+
+    public var configuration = SeverityConfiguration(.warning)
+
+    public init() {}
+
+    private static let regularExpression = regex(
+        "catch" + // The catch keyword
+        "(?:"   + // Start of the first non-capturing group
+        "\\s+"  + // At least one any type of whitespace character
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/UntypedErrorRule.swift]: A whitespace is not required if parentheses are used: `catch(let error) {`. 

Also, the rule currently wouldn't catch when they're used even with spaces.

File: Source/SwiftLintFramework/Rules/UntypedErrorRule.swift [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+                         kind: StatementKind,
+                         dictionary: [String: SourceKitRepresentable]) -> [StyleViolation] {
+
+        guard kind == .brace,
+            let offset = dictionary.offset else {
+                return []
+        }
+
+        return file.lines
+            .lazy
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/UntypedErrorRule.swift]: Does this really improve performance? I think `lines` will be calculated anyways.

File: Source/SwiftLintFramework/Rules/UntypedErrorRule.swift [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+            .filter { $0.byteRange.contains(offset) }
+            .reduce([]) { _, line in
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/UntypedErrorRule.swift]: wouldn't it be more semantic to use `flatMap` here?

File: Source/SwiftLintFramework/Rules/UntypedErrorRule.swift [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        "(?:"   + // Start of the second non-capturing group
+        "\\s*"  + // Zero or unlimited any whitespace character
+        ")"     + // End of the second non-capturing group
+        "\\{"     // Start scope character
+    )
+
+    public static let description = RuleDescription(
+        identifier: "untyped_error",
+        name: "Untyped error rule",
+        description: "Catch statements should not declare error variables without type casting",
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/UntypedErrorRule.swift]: Can you add a `.` in the end?

File: Source/SwiftLintFramework/Rules/UntypedErrorRule.swift [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        kind: .lint,
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/UntypedErrorRule.swift]: can you change to `.idiomatic`? `.lint` is more for validating things that could indicate a programmer's error.

File: Source/SwiftLintFramework/Rules/UntypedErrorRule.swift [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+        nonTriggeringExamples: [
+            "do {\n try foo() \n} catch {\n}",
+            "do {\n try foo() \n} catch Error.invalidOperation {\n} catch {\n}",
+            "do {\n try foo() \n} catch let error as? MyError {\n} catch {\n}",
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/UntypedErrorRule.swift]: this is not a major issue, but I think `catch let error as MyError` should be used. I got a warning while compiling with the optional cast:

```
 warning: cast from '_' to unrelated type 'MyError' always fails
```

File: Source/SwiftLintFramework/Rules/UntypedErrorRule.swift [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+                let range = NSRange(location: 0, length: line.content.bridge().length)
+                return type(of: self).regularExpression.matches(in: line.content, options: [], range: range)
+                    .map { match -> StyleViolation in
+                        let characterOffset = line.range.lowerBound + match.range.lowerBound
+                        let location = Location(file: file,
+                                                characterOffset: characterOffset)
+                        return StyleViolation(ruleDescription: type(of: self).description,
+                                              severity: configuration.severity,
+                                              location: location,
+                                              reason: configuration.consoleDescription)
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/UntypedErrorRule.swift]: Is there a reason why you're iterating through lines?

I have the impression that it'd be faster to make this a regular rule (not `ASTRule`) and just match the regex on the whole file, validating some syntax kinds (e.g. that `catch` is a `keyword`, as well as `let`/`var`).

Comment (dirtydanee) [reply Source/SwiftLintFramework/Rules/UntypedErrorRule.swift]: @marcelofabri Thanks for the review! This suggestion was a really spot on. I have decided to use ASTRule to be able to solve the issue with the commented out code. However the syntax kind validation works great, simplified the code by tons and should be faster too.

File: Source/SwiftLintFramework/Rules/UntypedErrorRule.swift (hunk 1 above)
Comment (dirtydanee) [reply Source/SwiftLintFramework/Rules/UntypedErrorRule.swift]: Updated the naming as you suggested, its for sure more clear what does it.

File: Source/SwiftLintFramework/Rules/UntypedErrorRule.swift (hunk 2 above)
Comment (dirtydanee) [reply Source/SwiftLintFramework/Rules/UntypedErrorRule.swift]: Fixed the indentation and removed the `\n` from the `catch` statements scope.

File: Source/SwiftLintFramework/Rules/UntypedErrorRule.swift (hunk 3 above)
Comment (dirtydanee) [reply Source/SwiftLintFramework/Rules/UntypedErrorRule.swift]: Updated the pattern to capture the errors with parentheses with and without spaces.

**Comment by dirtydanee** [comment]
@marcelofabri thoughts on the update?

**Comment by marcelofabri** [comment]
@dirtydanee Merged in #2101 🎉

**Pull request closed without merging**